Splitview

Wouter01 opened the pull request:
# Description

This PR adds splitview support for the editor. It includes the following features:

- Ability to split editor vertically or horizontally, by right clicking a tab or by pressing the split button in the tab view.
- Translucency for the toolbar
- Translucency for the statusbar
- Improvements to terminal dragging
- Forward / Backward funcionality for each TabGroup
- Closing a TabGroup by pressing the X button
- Probably some other things I forgot

There are also some regressions, which will be addressed in a later PR:
- Terminal state is not saved
- Tab state is not saved
- TabGroup state is not saved

# Related Issue

#338 
#1109 
# Checklist

- [x] Multiple Editors
- [x] Horizontal Split View
- [x] Vertical Split View
- [x] Nested Split View
- [x] Create Split View by right clicking on tab
- [ ] ~~Create Split View by dragging file into editor~~
- [ ] ~~Move tab to other editor by dragging~~
- [x] Make top editors go under toolbar
- [x] Make bottom editors go under bottom toolbar
- [x] Adjustments to tabs
- [x] Refactor `WorkspaceDocument` and others to support multiple tabgroups
- [x] Fix UI issues
- [x] implement close / backward / forward buttons
- [x] Add support for temporary tabs
- [x] Add menu when long-pressing forward/backward icon 

<!--- Add things that are not yet implemented above -->
- [x] I read and understood the [contributing guide](https://github.com/CodeEditApp/CodeEdit/blob/main/CONTRIBUTING.md) as well as the [code of conduct](https://github.com/CodeEditApp/CodeEdit/blob/main/CODE_OF_CONDUCT.md)
- [x] My changes generate no new warnings
- [x] My code builds and runs on my machine
- [x] I documented my code
- [x] Review requested

# Screenshots

<!--- REQUIRED: if issue is UI related -->

<!--- IMPORTANT: Fill out all required fields. Otherwise we might close this PR temporarily -->

iggy890 commented:
Nice job!

austincondiff commented:
Again, fantastic work. Here is a list of things I have found in my testing.

I can't scroll the tab bar, when opening other editors, sometimes it works in some editors but not in all.

https://user-images.githubusercontent.com/806104/222984235-dac3a2ed-c5ff-4955-b9f2-13016fdcbd9f.mov

The active tab is in _italics_, only temporary tabs should be.

https://user-images.githubusercontent.com/806104/222984252-f8e7c7bc-0b7b-4c48-b15a-ec97f5a8e553.mov

I'd remove the extra animations
  - When the tab closes
  - The split symbol button when holding the option key

https://user-images.githubusercontent.com/806104/222984344-002233db-eeb3-4e11-b894-d62514b95987.mov

Upon clicking the close button on a tab, the next tab's close button does not become visible until moving the mouse making it difficult to close multiple tabs quick. I've seen this happen most of the time, sometimes it works.

https://user-images.githubusercontent.com/806104/222984718-a3248c89-3166-407e-9d60-3a35a7611cd4.mov

I have to click in the editor in order to then click the split button, otherwise it cannot be clicked. Clicking the split button of an editor not in focus does not work no matter how many times it is clicked.

https://user-images.githubusercontent.com/806104/222985046-463a44ff-39e2-4c47-b85d-965e1605e04b.mov

When closing a vertical split, it just removes the containing view and not the split itself.

https://user-images.githubusercontent.com/806104/222985196-89375161-1e7c-4f5f-8bd5-7f9bd5e33660.mov

austincondiff reviewed (changes requested):
Again, fantastic work. Here are several things I have found in my testing.

https://user-images.githubusercontent.com/806104/222984235-dac3a2ed-c5ff-4955-b9f2-13016fdcbd9f.mov

I can't scroll the tab bar, when opening other editors, sometimes it works in some editors but not in all.

https://user-images.githubusercontent.com/806104/222984252-f8e7c7bc-0b7b-4c48-b15a-ec97f5a8e553.mov

The active tab is in _italics_, only temporary tabs should be.

https://user-images.githubusercontent.com/806104/222984344-002233db-eeb3-4e11-b894-d62514b95987.mov

I'd remove the extra animations
  - When the tab closes
  - The split symbol button when holding the option key

https://user-images.githubusercontent.com/806104/222984718-a3248c89-3166-407e-9d60-3a35a7611cd4.mov

Upon clicking the close button on a tab, the next tab's close button does not become visible until moving the mouse making it difficult to close multiple tabs quick. I've seen this happen most of the time, sometimes it works.

https://user-images.githubusercontent.com/806104/222985046-463a44ff-39e2-4c47-b85d-965e1605e04b.mov

I have to click in the editor in order to then click the split button, otherwise it cannot be clicked. Clicking the split button of an editor not in focus does not work no matter how many times it is clicked.

https://user-images.githubusercontent.com/806104/222985196-89375161-1e7c-4f5f-8bd5-7f9bd5e33660.mov

When closing a horizontal split that contains vertical splits, it just removes the containing view and not the split itself. Below is another example of this...

https://user-images.githubusercontent.com/806104/222985728-875e957b-fa5c-46e3-9ab1-df92a4989587.mp4

As you can see, I am also having trouble getting the editor split to focus. It requires that I click on the active tab in the split to focus the editor.

That is all I have found so far. If I see anything else I will add them here.

Wouter01 commented:
Thanks for the extensive review @austincondiff! I've addressed some of the issues:

> I can't scroll the tab bar, when opening other editors, sometimes it works in some editors but not in all.

This should be fixed.

 
> The active tab is in _italics_, only temporary tabs should be.

I'm looking further into this, I might need to change how I represent a temporary tab.

> I'd remove the extra animations
> 
> * When the tab closes
> * The split symbol button when holding the option key

I'll try to add these under a toggle "increase motion"


> Upon clicking the close button on a tab, the next tab's close button does not become visible until moving the mouse making it difficult to close multiple tabs quick. I've seen this happen most of the time, sometimes it works.

This also appears in the main branch and is an issue with the tabbarview, it should not be resolved here.

> Clicking the split button of an editor not in focus does not work no matter how many times it is clicked.

Fixed.


> When closing a horizontal split that contains vertical splits, it just removes the containing view and not the split itself. Below is another example of this...

I'll need to do some digging here.

> As you can see, I am also having trouble getting the editor split to focus. It requires that I click on the active tab in the split to focus the editor.
 I wasn't able to reproduce this issue unfortunately.

matthijseikelenboom commented on CodeEdit/Features/NavigatorSidebar/ExtensionNavigator/ExtensionNavigatorItemView.swift:
```diff
@@ -13,7 +13,7 @@ struct ExtensionNavigatorItemView: View {
     var body: some View {
         Button {
-            document.openTab(item: plugin)
+//            document.openTab(item: plugin)
```
Maybe this button can be removed?

matthijseikelenboom commented on CodeEdit/Features/NavigatorSidebar/ProjectNavigator/OutlineView/OutlineView.swift:
```diff
@@ -17,6 +17,9 @@ struct OutlineView: NSViewControllerRepresentable {
     var prefs: AppPreferencesModel = .shared
 
+    // This is mainly just used to trigger a view update.
+    @Binding var selection: WorkspaceClient.FileItem?
```
Break after attribute definition

matthijseikelenboom commented on CodeEdit/Features/SplitView/EditorView.swift:
```diff
@@ -0,0 +1,93 @@
+
+import SwiftUI
+
+struct EditorView: View {
```
Multiple cases of breaks after attribute definition in this file

matthijseikelenboom commented on CodeEdit/Features/SplitView/SplitView.swift:
```diff
@@ -0,0 +1,29 @@
+struct SplitView<Content: View>: View {
+    var content: Content
+
+    @State var viewController: SplitViewController
```
Break after attribute definition

matthijseikelenboom commented on CodeEdit/Features/SplitView/SplitViewControllerView.swift:
```diff
@@ -0,0 +1,89 @@
+    var viewController: SplitViewController
+
+    func makeNSViewController(context: Context) -> SplitViewController {
+//        viewController.items = children.map { SplitViewItem(child: $0) }
```
Maybe remove this line?

matthijseikelenboom commented on CodeEdit/Features/SplitView/SplitViewControllerView.swift:
```diff
@@ -0,0 +1,89 @@
+            let numerator = splitView.isVertical ? splitView.frame.width : splitView.frame.height
+            for idx in 0..<controller.items.count-1 {
+                // If the next view is collapsed, don't reposition the divider.
+                guard !controller.items[idx+1].item.isCollapsed else { continue }
```
This is a complex line, perhaps wrap it in a descriptive `private func`

matthijseikelenboom commented on CodeEdit/Features/SplitView/SplitViewControllerView.swift:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    override func viewDidLoad() {
+//        splitView.arrangesAllSubviews = false
```
Maybe remove this line?

matthijseikelenboom commented on CodeEdit/Features/SplitView/SplitViewData.swift:
```diff
@@ -0,0 +1,67 @@
+import SwiftUI
+
+class SplitViewData: ObservableObject {
+    @Published var tabgroups: [TabGroup]
```
Break after attribute definition

lukepistrol commented:
I'll have a look at this tomorrow 😊

matthijseikelenboom commented on CodeEdit/Features/StatusBar/Views/StatusBarItems/StatusBarToggleDrawerButton.swift:
```diff
@@ -11,7 +11,10 @@ internal struct StatusBarToggleDrawerButton: View {
     private var model: StatusBarViewModel
 
-    init() {
+    @Binding var collapsed: Bool
```
Break after attribute definition

matthijseikelenboom commented on CodeEdit/Features/StatusBar/Views/StatusBarView.swift:
```diff
@@ -23,75 +23,79 @@ struct StatusBarView: View {
+
+    var proxy: SplitViewProxy
+
+    @Binding var collapsed: Bool
```
Break after attribute definition

matthijseikelenboom commented on CodeEdit/Features/Tabs/Models/TabManager.swift:
```diff
@@ -0,0 +1,47 @@
+
+class TabManager: ObservableObject {
+    /// Collection of all the tabgroups.
+    @Published var tabGroups: TabGroup
+
+    /// The TabGroup with active focus.
+    @Published var activeTabGroup: TabGroupData {
```
Break after attribute definition

matthijseikelenboom commented on CodeEdit/Features/Tabs/TabGroup/TabGroupData.swift:
```diff
@@ -0,0 +1,205 @@
+import OrderedCollections
+import DequeModule
+
+final class TabGroupData: ObservableObject, Identifiable {
```
Multi cases of break after attribute definition

matthijseikelenboom commented on CodeEdit/Features/Tabs/TabGroup/WorkspaceTabGroupView.swift:
```diff
@@ -0,0 +1,76 @@
+
+import SwiftUI
+
+struct WorkspaceTabGroupView: View {
```
Multiple cases of break after attribute definition

matthijseikelenboom commented on CodeEdit/Features/Tabs/TabGroup/WorkspaceTabGroupView.swift:
```diff
@@ -0,0 +1,76 @@
+            .background(EffectView(.titlebar, blendingMode: .withinWindow, emphasized: false))
+        }
+        .focused($focus, equals: tabgroup)
+//        .focused($isFocused)
+//        .onChange(of: isFocused) { focused in
+//            print("Focus change for \(tabgroup.selected?.fileName)", focused)
+//            if focused {
+//                tabManager.activeTabGroup = tabgroup
+//            }
+//        }
```
Maybe delete this?

matthijseikelenboom commented on CodeEdit/Features/Tabs/Views/TabBarItemView.swift:
```diff
@@ -8,15 +8,23 @@
     @Environment(\.isFullscreen)
     private var isFullscreen
 
+    @EnvironmentObject var tabManager: TabManager
```
Break after attribute definition

matthijseikelenboom commented on CodeEdit/WindowObserver.swift:
```diff
@@ -22,8 +22,14 @@ struct WindowObserver<Content: View>: View {
     @StateObject
     private var prefs: AppPreferencesModel = .shared
 
+    @State var modifierFlags: NSEvent.ModifierFlags = []
```
Break after attribute definition

matthijseikelenboom commented on CodeEdit/WorkspaceView.swift:
```diff
@@ -9,16 +9,15 @@ import SwiftUI
 import AppKit
 
 struct WorkspaceView: View {
```
Multiple cases of break after attribute definition

matthijseikelenboom reviewed (changes requested):
So, here is my first review, I haven't ran this branch locally yet, just went through the code. The main this overall critique is that no function has Swift DocC, which perhaps is a lot of work to add, but it should be done in my opinion.

Wouter01 commented:
> So, here is my first review, I haven't ran this branch locally yet, just went through the code. The main this overall critique is that no function has Swift DocC, which perhaps is a lot of work to add, but it should be done in my opinion.

Thanks for the feedback! I've added DocC comments to all (most?) functions I added. I've also updated some style changes and other suggestions you made.

bombardier200 commented:
@Wouter01 was just testing this out, it seems that the bottom drawer no longer allows you to select Debugger or Output, not for sure if this on purpose or not!

Wouter01 commented:
> @Wouter01 was just testing this out, it seems that the bottom drawer no longer allows you to select Debugger or Output, not for sure if this on purpose or not!

Must've messed something up there 😅, I'll have a look

Wouter01 commented:
> @Wouter01 was just testing this out, it seems that the bottom drawer no longer allows you to select Debugger or Output, not for sure if this on purpose or not!

Looks like there was no code for those yet? There was however an unused switch, which I probably removed

austincondiff reviewed (dismissed):
After thorough testing, I cannot reproduce any of my previously reported bugs! 👏

Way to go @Wouter01, great work. What an accomplishment! 🙌

austincondiff commented:
@Wouter01 One quick thing I just found, after clicking the back and forward arrows of an inactive split, it doesn't bring the inactive split into focus.

thecoolwinter reviewed (changes requested):
A couple things beyond the comments I've left:
- There's a regression in the native tab style. The tab background is the active tab color and tabs consume the entire tab area when opened. See screenshots:
Before:
<img width="986" alt="Screenshot 2023-03-08 at 12 56 15 PM" src="https://user-images.githubusercontent.com/35942988/223810114-20bbce4d-9706-48a4-844b-3e176d0e663e.png">
After:
<img width="1001" alt="Screenshot 2023-03-08 at 12 58 36 PM" src="https://user-images.githubusercontent.com/35942988/223810247-4edc58d5-32f0-43ae-89c7-aa146d77cd22.png">
The background color also is the active color if you scroll to the left or right.

- After clicking the back and forward arrows of an inactive split, it doesn't bring the inactive split into focus. This looks like it's a quick fix by just setting `tabManager.activeTabGroup` in those button handlers.

austincondiff commented:
> A couple things beyond the comments I've left:
> 
> * There's a regression in the native tab style. The tab background is the active tab color and tabs consume the entire tab area when opened. See screenshots:
>   Before:
> 
> <img alt="Screenshot 2023-03-08 at 12 56 15 PM" width="986" src="https://user-images.githubusercontent.com/35942988/223810114-20bbce4d-9706-48a4-844b-3e176d0e663e.png">
> 
> After:
> <img alt="Screenshot 2023-03-08 at 12 58 36 PM" width="1001" src="https://user-images.githubusercontent.com/35942988/223810247-4edc58d5-32f0-43ae-89c7-aa146d77cd22.png">
> 
> The background color also is the active color if you scroll to the left or right.
> * After clicking the back and forward arrows of an inactive split, it doesn't bring the inactive split into focus. This looks like it's a quick fix by just setting `tabManager.activeTabGroup` in those button handlers.

@thecoolwinter I actually don't mind tabs taking up the full width, this is how they work in Finder and every other app that supports native tabs. However, there is a slight problem, when there are no tabs, it looks like there is one big empty tab in the tab bar. It should instead be the darker background color behind the buttons. 

<img width="1112" alt="image" src="https://user-images.githubusercontent.com/806104/223832754-830e830b-94ae-420f-b4f7-136aded16ed3.png">

austincondiff commented:
@Wouter01 One additional thing I just found, after clicking the back and forward arrows of an inactive split, it doesn't bring the inactive split into focus.

matthijseikelenboom reviewed (approved):
Looks good to me, just fix @thecoolwinter 's comments and then it can be merged I think.

Wouter01 commented on CodeEdit/Features/Tabs/Views/TabBarView.swift:
```diff
@@ -413,23 +438,87 @@ struct TabBarView: View {
+        .buttonStyle(.plain)
         .padding(.horizontal, 7)
-        .opacity(activeState != .inactive ? 1.0 : 0.5)
+        .brightness(activeState == .inactive ? -0.3 : 0)
```
This does indeed work correctly, and will make the view darker. However, I've switched back to using opacity as this change wasn't needed anymore

Wouter01 commented on CodeEdit/Features/InspectorSidebar/InspectorSidebarView.swift:
```diff
@@ -21,28 +24,22 @@ struct InspectorSidebarView: View {
+                default: EmptyView()
                 }
-            } else {
-                NoSelectionInspectorView()
```
Nope, I must've done this by accident. Fixed it

Wouter01 commented:
> @Wouter01 One additional thing I just found, after clicking the back and forward arrows of an inactive split, it doesn't bring the inactive split into focus.

@thecoolwinter mentioned that too, it's fixed!

austincondiff reviewed (approved):
Approved, for the last time! After that I am not approving again... 😜

thecoolwinter reviewed (approved):
Looks good. This is a lot of work nice job!
